sw/host/opentitanlib/src/transport/hyperdebug/dfu.rs

```diff
@@ -0,0 +1,433 @@
+    if let Some(fw) = OFFICIAL_FIRMWARE {
+        Ok(Some(get_hyperdebug_firmware_version(fw)?))
+    } else {
+        Ok(None)
+    }
+}
+
+/// Helper method to verify that the given binary image looks like a HyperDebug firmware image.
+fn validate_firmware_image(firmware: &[u8]) -> Result<()> {
+    get_hyperdebug_firmware_version(firmware)?;
+    Ok(())
+}
+
+const EC_COOKIE: [u8; 4] = [0x99, 0x88, 0x77, 0xce];
+const EC_FIRMWARE_NAME_LEN: usize = 32;
+
+fn get_hyperdebug_firmware_version(firmware: &[u8]) -> Result<&str> {
+    let Some(pos) = firmware[0..1024]
+        .chunks(4)
+        .position(|c| c[0..4] == EC_COOKIE) else {
+            bail!(TransportError::FirmwareProgramFailed(
+                "File is not a HyperDebug firmware image".to_string()
+            ));
+        };
+    let firmware_name_field = &firmware[(pos + 1) * 4..(pos + 1) * 4 + EC_FIRMWARE_NAME_LEN];
+    let end = firmware_name_field
+        .iter()
+        .rev()
+        .position(|b| *b != 0x00)
+        .map(|j| EC_FIRMWARE_NAME_LEN - j)
+        .unwrap_or(0);
+    Ok(std::str::from_utf8(&firmware_name_field[0..end])?)
+}
+
+/// Helper method to perform flash programming using ST's DfuSe variant of the DFU protocol.
+/// This method is used both by the `Hyperdebug` and the `HyperdebugDfu` structs.
+pub fn update_firmware(
+    usb_device: &mut UsbBackend,
+    firmware: &Option<Vec<u8>>,
+    progress: &Option<Box<dyn Fn(Progress)>>,
+) -> Result<Option<Box<dyn Annotate>>> {
+    let firmware: &[u8] = if let Some(vec) = firmware.as_ref() {
+        validate_firmware_image(vec)?;
+        vec
+    } else {
+        OFFICIAL_FIRMWARE.unwrap()
+    };
+
+    let dfu_desc = scan_usb_descriptor(usb_device)?;
+
+    // Exclusively claim DFU interface, preparing for control requests.
+    usb_device.claim_interface(dfu_desc.dfu_interface)?;
+
+    if wait_for_idle(usb_device, dfu_desc.dfu_interface)? != DFU_STATE_APP_IDLE {
+        // Device is already running DFU bootloader, proceed to firmware transfer.
+        do_update_firmware(usb_device, dfu_desc, firmware, progress)?;
+        return Ok(None);
+    }
+
+    // Device is running the HyperDebug firmware, not DFU bootloader.  Ask for switch to
+    // bootloader, and then restablish USB connection.  Switching will cause loss of USB
+    // connection, hence the function block to continue even if either `write_control()` or
+    // `wait_for_idle()` terminates with an `Err()`.
+    let _ = (|| -> Result<()> {
+        log::info!("Requesting switch to DFU mode...");
+        usb_device.write_control(
+            rusb::request_type(
+                rusb::Direction::Out,
+                rusb::RequestType::Class,
+                rusb::Recipient::Interface,
+            ),
+            USB_DFU_DETACH,
+            1000,
+            dfu_desc.dfu_interface as u16,
+            &[],
+        )?;
+        wait_for_idle(usb_device, dfu_desc.dfu_interface)?;
+        Ok(())
+    })();
```

Comment:
What is the reason for this javascript-like [IIFE](https://developer.mozilla.org/en-US/docs/Glossary/IIFE)?

Reply:
I tried explaining that in the comment just above the function:
> Switching will cause loss of USB connection, hence the function block to continue even if either `write_control()` or `wait_for_idle()` terminates with an `Err()`.

Reply:
I think some of the chaining methods of `Result` may be more appropriate here:

```
let _ = usb_device.write_control(...).and_then(|| wait_for_idle(...));
```

As I recall, IIFEs were always a scoping hack in Javascript.  In rust, we can use `_` in the let-binding to ignore the result, and use `Result`-chaining to achieve the desired flow control.

WDYT?

Reply:
I was not familiar with `and_then()`, I think avoiding the anonymous function makes the intent more clear.